pkg/frontend/querymiddleware/roundtrip.go

```diff
@@ -445,6 +445,34 @@ func newQueryMiddlewares(
 		)
 	}
 
+	// Does not apply to remote read as those are executed remotely and the enabling of PromQL experimental
+	// functions for those are not controlled here.
+	experimentalFunctionsMiddleware := newExperimentalFunctionsMiddleware(limits, log)
+	queryRangeMiddleware = append(
+		queryRangeMiddleware,
+		newInstrumentMiddleware("experimental_functions", metrics),
+		experimentalFunctionsMiddleware,
+	)
+	queryInstantMiddleware = append(
+		queryInstantMiddleware,
+		newInstrumentMiddleware("experimental_functions", metrics),
+		experimentalFunctionsMiddleware,
+	)
+
+	if cfg.PrunedQueries {
+		pruneMiddleware := newPruneMiddleware(log)
+		queryRangeMiddleware = append(
+			queryRangeMiddleware,
+			newInstrumentMiddleware("pruning", metrics),
+			pruneMiddleware,
+		)
+		queryInstantMiddleware = append(
+			queryInstantMiddleware,
+			newInstrumentMiddleware("pruning", metrics),
+			pruneMiddleware,
+		)
+	}
```

Comment:
Note to reviewers: I can see pros and cons to moving this immediately after the splitting and caching middleware.

Pros: if the entire result has already been cached, then we'll skip the work of pruning the expression

Cons: if at least two parts of the query haven't been cached, we'll end up applying pruning to the same expression multiple times

Reply:
>Why two parts? Isn't one uncached extent enough?

If the query is split into multiple extents, then the split query for each extent passes through all of middlewares after the cache/split middleware independently. So if multiple extents aren't cached, and the pruning middleware runs after the cache/split middleware, then pruning would be applied to each extent independently.

Reply:
So you meant twice within the same HTTP request. I think I understood twice within different HTTP requests. So that's where the confusion came from. Thanks for explaining